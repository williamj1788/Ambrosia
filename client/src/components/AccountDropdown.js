import React from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { clearUser } from '../redux/action';
import { GoogleLogout } from 'react-google-login';
import s from '../styles/AccountDropdown.module.scss';

export class AccountDropdown extends React.Component{
    
    state = {
        redirect: null,
    }
    
    signOutUser = () => {
        return fetch('/api/user/signout', {
            credentials: 'include',
        }).then(() => {
            this.props.dispatch(clearUser());
        });
    }

    setRedirect = location => {
        this.setState({
            redirect: location,
        });
    }

    componentDidUpdate(){ // Set redirect back to null or less you get a infinate loop
        if(this.state.redirect){
            this.setState({
                redirect: null
            });
        }
    }
    
    render(){
        const { user, show } = this.props;
        const { redirect } = this.state;
        if(redirect){
            return <Redirect to={redirect} />
        }
        let style = {
            height: window.innerWidth > 1000 ? user.admin ? '200px' : '100px' : user.admin ? '100px': '50px'
        };
        if(!show){
            if(window.innerWidth > 1000){
                style.height = '0';
            }else{
                style.width = '0';
            }
        }
        if(window.innerWidth < 1000 && user.admin){
            style.top = "-40px";
        }

        return(
            <div className={s.dropDown} style={style}>
                <div className={s.content}>
                    {user.admin && <Tab onClick={() => this.setRedirect('/admin/metrics')} text="Metrics" />} 
                    {user.admin && <Tab onClick={() => this.setRedirect('/admin/products')} text="Product" />} 
                    <Tab onClick={() => this.setRedirect('/user/orders')} text="Order History" />
                    <GoogleLogout
                        onLogoutSuccess={() => {}}
                        render={
                            renderProps => (
                                <div id='signout' className={s.tab} onClick={renderProps.onClick}>
                                    <div 
                                    onClick={this.signOutUser}
                                    style={{
                                        width: '100%',
                                        height: '100%',
                                        display: 'flex',
                                        justifyContent: 'center',
                                        alignItems: 'center'
                                    }} >
                                        <span>Sign Out</span>
                                    </div>
                                </div>
                            )
                        }
                    />
                </div>
            </div>
        )
    }
}

export const Tab = ({onClick, text}) => {
    return(
        <div className={s.tab} onClick={onClick} >
            <span>{text}</span>
        </div>
    )
}

export default connect(state => {
    return {user: state.user}
})(AccountDropdown);