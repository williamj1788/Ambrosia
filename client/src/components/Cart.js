import React from 'react';
import s from '../styles/Cart.module.scss';
import { connect } from 'react-redux';
import { editOrder, removeOrder } from '../redux/action';
import { FaTimes } from 'react-icons/fa';
import uuid from 'uuid';
import { Redirect, withRouter } from 'react-router-dom';

class Cart extends React.Component{
    
    state = {
        active: 1,
        redirect: false,
        error: null,
    }

    handleClick = () => {
        if(!this.props.user){
            this.setRedirect('/signup');
        }else if(this.props.orders.length <= 0){
            this.setState({
                error: 'There are no items'
            });
        }else{
            this.setRedirect('/order');
        }
    }

    removeOrder = id => {
        this.props.dispatch(removeOrder(id));
    }

    editOrder = (id, qty) => {
        this.props.dispatch(editOrder({id,qty}));
    }


    setActive = value => {
        this.setState({active: value});
    }

    setRedirect = value => {
        this.setState({redirect: value});
    }
    
    render(){
        const { redirect, active, error } = this.state;
        const { orders, location, toggle } = this.props;
        if(redirect){
            if(redirect === '/signup' && location.pathname !== '/signup'){
                return <Redirect push to={redirect} />
            }else if(redirect === '/order' && location.pathname !== '/order'){
                return <Redirect push to={redirect} />
            }else{
                toggle();
            }
        }

        return(
            <div className={s.dark}>
                <div className={s.cart}>
                    <Header toggle={toggle} />
                    <OrderContainer 
                    orders={this.props.orders} 
                    remove={this.removeOrder}
                    edit={this.editOrder}
                    active={active}
                    setActive={this.setActive}  />
                    <Footer orders={orders} onClick={this.handleClick} error={error} />
                </div>
            </div>
        )
    }
}

const Header = ({ toggle }) => {
    return(
        <div className={s.header} >
            <span>Cart</span>
            <div className={s.close} onClick={toggle} >
                <FaTimes size='2em' />
            </div>
        </div>
    )
}

const Footer = ({ orders, onClick, error }) => {
    let subTotal = orders.reduce((acc, val) => {
        return acc + (val.price * val.qty);
    }, 0);

    let tax = subTotal * 0.06;

    let deliveryFee = 3;

    let total = subTotal + tax + deliveryFee;

    return(
        <div className={s.footer}>
            <div className={s.flexTotal}>
                <span>Sub Total:</span>
                <span>{'$' + subTotal.toFixed(2)}</span>
            </div>
            <div className={s.flexTotal}>
                <span>Tax(6%):</span>
                <span>{'$' + tax.toFixed(2)}</span>
            </div>
            <div className={s.flexTotal}>
                <span>Delivery fees:</span>
                <span>{'$' + deliveryFee.toFixed(2)}</span>
            </div>
            <div className={s.flexTotal}>
                <span>Total:</span>
                <span>{'$' + total.toFixed(2)}</span>
            </div>
            <button onClick={onClick} type='button' className={s.button} >Checkout</button>
            {error && <p className={s.error}>{error}</p>}
        </div>
    )
}

const OrderContainer = ({ orders, remove, edit, active, setActive }) => {
    const MaxOrders = 7;
    let  curOrders = orders.slice((MaxOrders * (active - 1)), (MaxOrders * active));
    curOrders = curOrders.map((order) => {
        return <Order key={order.id} remove={remove} edit={edit} {...order} />
    });

    if(curOrders.length < 1 && active !== 1){
        setActive(active - 1);
    }

    let counters = [];
    for(let i = 1; i <= Math.ceil(orders.length / MaxOrders); i++){
        counters.push(<button key={uuid()} onClick={() => setActive(i)} className={`${s.counter} ${active === i && s.active}`}>{i}</button>);
    };
    
    return(
        <div className={s.OrderContainer}>
            {curOrders}
            <div className={s.counterCounter}>
                {counters.length > 1 && counters}
            </div>
        </div>
    )
};

const Order = ({ name, price, qty, id, remove, edit }) => {
    return(
        <div className={s.order}>
            <div className={s.orderInfo}>
                <p className={s.orderTitle}>{name}</p>
                <p className={s.orderPrice}>{'$' + price.toFixed(2)}</p>
                <select className={s.select} onChange={event => edit(id, event.target.value)} defaultValue={qty} name="Qty">
                    <option value="1">1</option>
                    <option value="2">2</option>
                    <option value="3">3</option>
                    <option value="4">4</option>
                    <option value="5">5</option>
                    <option value="6">6</option>
                    <option value="7">7</option>
                </select>
            </div>
            <button type='button' className={s.orderRemove} onClick={() => remove(id)} >Remove</button>
        </div>
    )
}

export default withRouter(connect(state => {
    return{
        user: state.user,
        orders: state.orders
    }
})(Cart));