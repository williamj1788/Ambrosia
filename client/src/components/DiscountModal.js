import React from 'react';
import s from '../styles/Products.module.scss';
import { FaTimes } from "react-icons/fa";

class DiscountModal extends React.Component{
    render(){
        return(
            <div className={s.dark}>
                <div className={s.discountModal}>
                    <div className={s.header} style={{backgroundColor: '#0033ff'}} >
                        <span style={{color: 'white'}} >Add A Discount</span>
                        <div className={s.close} onClick={this.props.show}>
                            <FaTimes size="1.75em" />
                        </div>
                    </div>
                    <p className={s.discountTitle}>Product - Pizza</p>
                    <div className={s.discountFormRecord}>
                        <p className={s.discountLabel}>Old Price:</p>
                        <p className={s.discountRignt}>12.99</p>
                    </div>
                    <form >
                        <div className={s.discountFormRecord}>
                            <label className={s.discountLabel} htmlFor="newPrice">New Price:</label>
                            <input className={s.discountRignt} type="number" name="newPrice" />
                        </div>
                        <div className={s.discountFormRecord}>
                            <label className={s.discountLabel} htmlFor="newPrice">Expire in:</label>
                            <select className={s.discountRignt} name="expireIn">
                                <option value="1">1 day</option>
                                <option value="2">2 days</option>
                                <option value="3">3 days</option>
                                <option value="4">4 days</option>
                                <option value="5">5 days</option>
                                <option value="6">6 days</option>
                                <option value="7">7 days</option>
                            </select>
                        </div>
                        <button className={s.discountButton} type='submit'>Add</button>
                    </form>
                    
                </div>
            </div>
        )
    }
}

export default DiscountModal;