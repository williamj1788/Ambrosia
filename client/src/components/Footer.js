import React from 'react';
import s from '../styles/Footer.module.scss';

import twitterIcon from '../images/twitter_icon.png';
import facebookIcon from '../images/Facebook-Icon.png';
import instagramIcon from '../images/instagram.png';

import locationIcon from '../images/location-icon.png';
import phoneIcon from '../images/phone-icon.png';
import emailIcon from '../images/email-icon.svg';

const Footer = () => {
    return(
        <footer className={s.footer}>
            <div className={s.container}>
                <div className={s.opening}>
                    <h4 className={s.title}>Opening Hours</h4>
                    <div className={s.hours}>
                        <div className={s.hoursSection}>
                            <p className={s.hoursTitle}>Monday - Friday</p>
                            <p className={s.hoursText}>8.00 am - 10.00 pm</p>
                        </div>
                        <div className={s.hoursSection}>
                            <p className={s.hoursTitle}>Saturday</p>
                            <p className={s.hoursText}>10.00 am - 5.00 pm</p>
                        </div>
                        <div className={s.hoursSection}>
                            <p className={s.hoursTitle}>Sunday</p>
                            <p className={s.hoursText}>Closed</p>
                        </div>
                    </div>
                </div>
                <div className={s.contact}>
                    <h4 className={s.title}>Contact Us</h4>
                    <div className={s.contactSocialIconContainer}>
                        <img className={s.contactSocialIcon} src={twitterIcon} alt="Social Media Icon"/>
                        <img className={s.contactSocialIcon} src={facebookIcon} alt="Social Media Icon"/>
                        <img className={s.contactSocialIcon} src={instagramIcon} alt="Social Media Icon"/>
                    </div>
                    <div className={s.contactSection}>
                        <img className={s.contactSectionIcon} src={locationIcon} alt="Location Icon"/>
                        <p className={s.contactSectionText}>203 Fake St. Mountain View, San Francisco, California, USA</p>
                    </div>
                    <div className={s.contactSection}>
                        <img className={s.contactSectionIcon} src={phoneIcon} alt="Phone Icon"/>
                        <p className={s.contactSectionText}>+1 978-236-7889</p>
                    </div>
                    <div className={s.contactSection}>
                        <img className={s.contactSectionIcon} src={emailIcon} alt="Email Icon"/>
                        <p className={s.contactSectionText}>JacquezWilliams@teleworm.us</p>
                    </div>
                </div>
            </div>
        </footer> 
    )
}

export default Footer;