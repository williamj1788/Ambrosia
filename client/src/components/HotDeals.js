import React from 'react';
import s from '../styles/HotDeals.module.scss';
import PizzaIcon from '../images/Pizza_icon_white.png';
import { connect } from 'react-redux';

export class HotDeals extends React.Component{

    findProductsWithDiscounts = () => {
        return this.props.products.filter(product => {
            return product.discountObj.length !== 0;
        })
    }
    
    render(){
        return(
            <section className={s.hotDeals}>
                <h3 className={s.hotDealsTitle}>Hot Deals</h3>
                <DealContainer products={this.findProductsWithDiscounts()} />
            </section>
        )
    }
}

export const DealContainer = ({ products }) => {
    const deals = products.map(product => {
        return <Deal {...product} />
    });


    return(
        <div className={s.dealsContainer}>
            {deals}
        </div>
    )
}


export const Deal = ({ name, picture, description, price, discountObj }) => {
    return(
        <div className={s.deal}>
            <img className={s.dealImg} src={picture} alt="product"/>
            <div className={s.dealInfo}>
                <span className={s.dealTitle}>{name}</span>
                <p className={s.dealDesc}>{description}</p>
                <div className={s.dealStat}>
                    <span className={s.dealPrice}>{getDiscountPercent(discountObj[0].price, price) + '% off'}</span>
                    <span className={s.dealPrice}>{Math.floor(1 + daysBetweenDates(discountObj[0].expiresAt, new Date())) + ' days left'}</span>
                </div>
                <div className={s.dealPriceContainer}>
                    <span className={s.dealPrice}><s>{'$' + price}</s></span>
                    <span className={s.dealPrice}>{'$' + discountObj[0].price}</span>
                    <button className={s.dealButton}>Place Order</button>
                </div>
            </div>
        </div>
    )
}

function roundNumber(num, target = 2){
    return Math.round(num * Math.pow(10, target)) / 100
};

function getDiscountPercent(discount, price) {
   return Math.round((1 - roundNumber(discount / price)) * 100)
}

function daysBetweenDates(day1, day2) {
    console.log(day1, day2)
    console.log(Math.round(((new Date(day1).getTime() - day2.getTime()) / (1000 * 60 * 60 * 24)) * 100) / 100)
    return Math.round(((new Date(day1).getTime() - day2.getTime()) / (1000 * 60 * 60 * 24)) * 100) / 100
}

export default connect(state => {
    return {
        products: state.products
    }
})(HotDeals);