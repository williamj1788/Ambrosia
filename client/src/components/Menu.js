import React from 'react';
import Navbar from './Navbar';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { setProducts } from '../redux/action';
import s from '../styles/Menu.module.scss';

import Product from './MenuProduct';

export class Menu extends React.Component{
    
    state = {
        loading: !this.props.products,
        redirect: false,
        productTarget: null,
    }

    redirectToProduct = productTarget => {
        sessionStorage.setItem('scroll',window.scrollY);
        this.setState({
            redirect: true,
            productTarget,
        });
    }

    componentDidMount(){
        if(!this.props.products){
            this.fetchProducts();
        }
    }

    componentDidUpdate(){
        window.scrollTo(0, parseInt(sessionStorage.getItem('scroll')));
        
        // set state back to default or else you get a infinite redirect loop
        if(this.state.redirect){
            this.setState({
                redirect: false,
                productTarget: null,
            });
        }
    }

    fetchProducts = () => {
        fetch('/api/admin/products')
        .then(res => res.json())
        .then(res => this.props.dispatch(setProducts(res)))
        .then(() => {
            this.setState({
                loading: false,
            });
        })
        .catch(console.log);
    }

    getProductsByType = type => {
        return this.props.products.filter(product => product.type === type);
    }



    render(){
        const { redirect, productTarget, loading } = this.state;
        if(redirect){ 
            return <Redirect to={`/menu/${productTarget}`} /> 
        }

        if(loading){
            return <div>Loading...</div>
        }
        return(
            <div>
                <Navbar />
                <div className={s.content}>
                    <h1 className={s.title}>Menu</h1>
                    <TabContainer redirect={this.redirectToProduct} productPage={this.props.match.params.product} >
                        <Tab product='pizza'>Pizza</Tab>
                        <Tab product='pasta'>Pasta</Tab>
                        <Tab product='bread'>Bread</Tab>
                        <Tab product='dessert'>Desserts</Tab>
                        <Tab product='drink'>Drinks</Tab>
                    </TabContainer>
                    <ProductContainer products={this.getProductsByType(this.props.match.params.product)} />
                </div>
            </div>
        )
    }
}

function giveChildrenProps(children, props) {
    return React.Children.map(children, child => {
        return React.cloneElement(child, props);
    });
}

const TabContainer = ({ redirect, productPage, children }) => {
    children = giveChildrenProps(children, { redirect, productPage });
    return(
        <div className={s.tabContainer}>
            {children}
        </div>
    )
}

const Tab = ({ productPage, redirect, product, children }) => {
    return(
        <button onClick={productPage !== product ? () =>{ redirect(product) } : undefined} className={`${s.tab} ${productPage === product ? s.active : ''}`}>{children}</button>
    )
} 

export const ProductContainer = ({ products, createOrder }) => {
    return(
        <div className={s.productContainer} >
            {products.map((product, index) => {
                return(
                    <Product key={index} {...product} />
                )
            })}
        </div>
    )
}

export default connect(state => {
    return{
        products: state.products,
        orders: state.orders,
    }
})(Menu);