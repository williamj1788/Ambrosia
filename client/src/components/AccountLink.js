import React from 'react';
import { connect } from 'react-redux';
import userIcon from '../images/UserIcon.png';
import UserArrow from '../images/Account-arrow.png';
import s from '../styles/Navbar.module.scss';
import { NavLinkWrapper } from './NavComponents';
import AccountDropdown from './AccountDropdown';

export class AccountLink extends React.Component{
    
    state = {
        showDropdown: false,
    }

    componentDidMount(){
        window.addEventListener('scroll', this.handleScroll);
    }

    handleScroll = () => {
        if(this.state.showDropdown){
            this.setState({
                showDropdown: false,
            })
        }
    }
    
    toggleDropdown = () =>{
        this.setState({
            showDropdown: !this.state.showDropdown
        });
    }

    render(){
        const { user } = this.props;
        const { showDropdown } = this.state
        if(user){
            return(
                <div style={{position: 'relative'}} >
                    <div className={s.navLink} style={{cursor: 'pointer'}} onClick={this.toggleDropdown} >
                        <img className={s.linkLogo} src={user.picture ? user.picture : userIcon} alt="User Icon" />
                        <span>{user.firstname}</span>
                        <img className={s.userArrow} src={UserArrow} alt="User Arrow"/>
                    </div>
                    <AccountDropdown show={showDropdown} />
                </div>
            )
        }
        return(
            <NavLinkWrapper to='/login' icon={userIcon}>Login</NavLinkWrapper>
        )
    }
}



export default connect(state => {
    return {user: state.user}
})(AccountLink);