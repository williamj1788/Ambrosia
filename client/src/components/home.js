import React from 'react';

import Navbar from './Navbar';
import ReviewSection from './ReviewSection';
import Footer from './Footer';
import HotDeals from './HotDeals';

import s from '../styles/Home.module.scss';

import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';

import PizzaIcon from '../images/Pizza_icon_white.png';
import MoneyIcon from '../images/Money_icon.png';
import TruckIcon from '../images/truck_icon.png';
import ArrowDown from '../images/arrow-down.png';


export class Home extends React.Component{
    
    state = {
        redirectToMenu: false
    }

    setRedirect = () => {
        this.setState({
            redirectToMenu: true,
        });
    }
    
    render(){
        if(this.state.redirectToMenu){ 
            return <Redirect to='/menu' />
        }
        return(
            <div>
                <Hero redirect={this.setRedirect}/>
                <About />
                <HotDeals />
                <ReviewSection />
                <Footer />
                <Copyright />
            </div>
        )
    }
}

const Hero = ({ redirect }) => {
    return(
        <section className={s.hero}>
            <div className={s.darken}>
                <Navbar />
                <p className={s.heroTitle}>Best Pizza in Town</p>
                <button onClick={redirect} className={s.heroButton}>View Menu</button>
                <img className={s.heroArrow} src={ArrowDown} alt="ArrowDown Icon"/>
            </div>
        </section>
    )
}

const About = () => {
    return(
        <section className={s.about}>
            <AboutTitle />
            <div className={s.flexContainer} >
                <Unit 
                img={PizzaIcon}
                title={'Great Pizza'}
                desc={'We go through multiple checks of each ingredient to ensure that our pizza is top quality'}
                />
                <Unit 
                img={MoneyIcon}
                title={'Low Prices'}
                desc={'We work with our local farmers to provide affordable and quality ingredients and pass the savings to our customers'}
                />
                <Unit 
                img={TruckIcon}
                title={'Fast Service'}
                desc={'With active drivers that are always on standby, we guarantee a delivery in under 30 mins or the order is free'}
                />
            </div>
        </section>
    )
}

const AboutTitle = () => {
    return(
        <div>
            <h2 className={s.aboutTitle}>About Us</h2>
            <p className={s.aboutDesc}>We are committed to providing our customers with excellent service that makes us deserving of being the best pizzeria in town</p> 
        </div>
    )
}

export const Unit = ({ img, title, desc }) => {
    return(
        <div className={s.AboutUnits}>
            <img className={s.aboutUnitImg} src={img} alt="Icon"/>
            <span className={s.aboutUnitTitle} >{title}</span>
            <p className={s.aboutUnitDesc}>{desc}</p>
        </div>
    )
}

const Copyright = () => {
    return(
        <div className={s.copyright}>
            <p className={s.copyrightText}>Copyright Â©2019 All rights reserved</p>
        </div>
    )
}

export default connect()(Home);