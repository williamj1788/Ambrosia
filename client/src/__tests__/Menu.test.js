import React from 'react';
import { shallow } from 'enzyme';
import { Menu } from '../components/Menu';

test('<Menu /> should render', () => {
    const props = {
        match: {
            params: {
                product: 'pizza'
            }
        }
    }
    
    const component = shallow(<Menu {...props} />);
    expect(component).toBeTruthy();
});
