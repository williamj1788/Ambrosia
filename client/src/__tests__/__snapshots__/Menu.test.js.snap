// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Tests Snapshot Test 1`] = `
<div
  className="product"
>
  <img
    alt="Product"
    className="productImg"
  />
  <div
    className="productInfo"
  >
    <p
      className="productName"
    >
      mock name
    </p>
    <p
      className="productDesc"
    >
      mock description
    </p>
    <div
      className="productOrder"
    >
      <span
        className="productPrice"
      />
      <form
        className="productForm"
      >
        <label
          className="productLabel"
          htmlFor="quantity"
        >
          Qty:
        </label>
        <select
          className="productSelect"
          defaultValue="1"
          name="quantity"
        >
          <option
            value="1"
          >
            1
          </option>
          <option
            value="2"
          >
            2
          </option>
          <option
            value="3"
          >
            3
          </option>
          <option
            value="4"
          >
            4
          </option>
          <option
            value="5"
          >
            5
          </option>
          <option
            value="6"
          >
            6
          </option>
          <option
            value="7"
          >
            7
          </option>
        </select>
        <button
          className="productButton"
          type="submit"
        >
          Place Order
        </button>
      </form>
    </div>
  </div>
</div>
`;
